dbaccess/block.go

```diff
@@ -0,0 +1,258 @@
+	db, err := ctx.db()
+	if err != nil {
+		return err
+	}
+
+	dbResult := db.
+		Model(&dbmodels.Block{}).
+		Where(&dbmodels.Block{AcceptingBlockID: rpcmodel.Uint64(currentAcceptingBlockID)}).
+		Update("accepting_block_id", newAcceptingBlockID)
+
+	dbErrors := dbResult.GetErrors()
+	if httpserverutils.HasDBError(dbErrors) {
+		return httpserverutils.NewErrorFromDBErrors("failed to update blocks: ", dbErrors)
+	}
+
+	return nil
+}
+
+// UpdateBlockAcceptingBlockID updates blocks with `blockID to be accepted by `acceptingBlockID `.
+// `acceptingBlockID` can be set nil.
+func UpdateBlockAcceptingBlockID(ctx Context, blockID uint64, acceptingBlockID *uint64) error {
+	db, err := ctx.db()
+	if err != nil {
+		return err
+	}
+
+	dbResult := db.
+		Model(&dbmodels.Block{}).
+		Where(&dbmodels.Block{ID: blockID}).
+		Update("accepting_block_id", acceptingBlockID)
+
+	dbErrors := dbResult.GetErrors()
+	if httpserverutils.HasDBError(dbErrors) {
+		return httpserverutils.NewErrorFromDBErrors("failed to update blocks: ", dbErrors)
+	}
+
+	return nil
+}
+
+// UpdateBlockIsChainBlock updates the block `blockID` by setting it's is_chain_block column to `isChainBlock`
+func UpdateBlockIsChainBlock(ctx Context, blockID uint64, isChainBlock bool) error {
+	db, err := ctx.db()
+	if err != nil {
+		return err
+	}
+
+	dbResult := db.
+		Model(&dbmodels.Block{}).
+		Where("id = ?", blockID).
+		Update("is_chain_block", isChainBlock)
+	dbErrors := dbResult.GetErrors()
+
+	if httpserverutils.HasDBError(dbErrors) {
+		return httpserverutils.NewErrorFromDBErrors("failed to update block: ", dbErrors)
+	}
+
+	return nil
+}
+
+// DoesBlockExist checks in the database whether a block with `blockHash` exists.
+func DoesBlockExist(ctx Context, blockHash string) (bool, error) {
+	db, err := ctx.db()
+	if err != nil {
+		return false, err
+	}
+
+	var blocksCount uint64
+	dbResult := db.
+		Model(&dbmodels.Block{}).
+		Where(&dbmodels.Block{BlockHash: blockHash}).
+		Count(&blocksCount)
+	dbErrors := dbResult.GetErrors()
+	if httpserverutils.HasDBError(dbErrors) {
+		return false, httpserverutils.NewErrorFromDBErrors("failed to find block: ", dbErrors)
+	}
+	return blocksCount > 0, nil
+}
+
+// BlocksCount returns the total number of blocks stored in the database
+func BlocksCount(ctx Context) (uint64, error) {
+	db, err := ctx.db()
+	if err != nil {
+		return 0, err
+	}
+
+	var count uint64
+	dbResult := db.Model(dbmodels.Block{}).Count(&count)
+	dbErrors := dbResult.GetErrors()
+	if httpserverutils.HasDBError(dbErrors) {
+		return 0, httpserverutils.NewErrorFromDBErrors("Some errors were encountered when counting blocks:", dbErrors)
+	}
+
+	return count, nil
+}
+
+// SelectedTipBlueScore returns the blue score of the selected tip
+func SelectedTipBlueScore(ctx Context) (uint64, error) {
+	db, err := ctx.db()
+	if err != nil {
+		return 0, err
+	}
+
+	var blueScore []uint64
+	dbResult := db.Model(&dbmodels.Block{}).
```

Comment:
> Can we somehow reuse logic from bluest block?
Not in a way that will greatly complicate both functions. I'll be happy to discuss this with you, we might be able to come up with something

> Can you please also consider using MAX?
Done

> Why do you use pluck instead of SELECT blue_score
Pluck is the gorm way to get a single column.

Reply:
> Can we somehow reuse logic from bluest block?
Not in a way that will greatly complicate both functions. I'll be happy to discuss this with you, we might be able to come up with something

> Can you please also consider using MAX?
Done

> Why do you use pluck instead of SELECT blue_score
Pluck is the gorm way to get a single column.